mlflow/tensorflow/__init__.py

```diff
@@ -1257,6 +1265,29 @@ def __restore_generator(prev_generator):
                 if log_models:
                     _log_keras_model(history, args)
 
+                if log_datasets:
+                    try:
+                        # create a CodeDatasetSource
+                        context_tags = context_registry.resolve_tags()
+                        source = CodeDatasetSource(
+                            mlflow_source_type=context_tags[MLFLOW_SOURCE_TYPE],
+                            mlflow_source_name=context_tags[MLFLOW_SOURCE_NAME],
+                        )
+                        training_data = kwargs["x"] if "x" in kwargs else args[0]
+                        validation_data = (
+                            kwargs["validation_data"] if "validation_data" in kwargs else None
+                        )
+                        _log_tensorflow_dataset(training_data, source, "train")
+                        if validation_data is not None:
+                            _log_tensorflow_dataset(validation_data, source, "eval")
```

Comment:
I ran the following code that specifies `validation_data` as a tuple of numpy arrays, but dataset logging failed with "unrecognized dataset type". Can we support this and add a test case?

```
"""Trains and evaluate a simple MLP
on the Reuters newswire topic classification task.
"""
import numpy as np
from tensorflow import keras
from tensorflow.keras.datasets import reuters
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Activation
from tensorflow.keras.preprocessing.text import Tokenizer

# The following import and function call are the only additions to code required
# to automatically log metrics and parameters to MLflow.
import mlflow

mlflow.tensorflow.autolog()

max_words = 1000
batch_size = 32
epochs = 5

print("Loading data...")
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=max_words, test_split=0.2)

print(len(x_train), "train sequences")
print(len(x_test), "test sequences")

num_classes = np.max(y_train) + 1
print(num_classes, "classes")

print("Vectorizing sequence data...")
tokenizer = Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode="binary")
x_test = tokenizer.sequences_to_matrix(x_test, mode="binary")
print("x_train shape:", x_train.shape)
print("x_test shape:", x_test.shape)

print("Convert class vector to binary class matrix (for use with categorical_crossentropy)")
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

print("Building model...")
model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation("relu"))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation("softmax"))

model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

history = model.fit(
    x_train, y_train, batch_size=batch_size, epochs=epochs, verbose=1, validation_split=0.1, validation_data=(x_train, y_train)
)
score = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=1)
print("Test score:", score[0])
print("Test accuracy:", score[1])
```

```
$ python train.py
2023-04-25 16:44:58.218156: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Loading data...
8982 train sequences
2246 test sequences
46 classes
Vectorizing sequence data...
x_train shape: (8982, 1000)
x_test shape: (2246, 1000)
Convert class vector to binary class matrix (for use with categorical_crossentropy)
y_train shape: (8982, 46)
y_test shape: (2246, 46)
Building model...
2023/04/25 16:45:08 INFO mlflow.utils.autologging_utils: Created MLflow autologging run with ID 'e680592ecc3a47a49ccfb735c6f0fc73', which will track hyperparameters, performance metrics, model artifacts, and lineage information for the current tensorflow workflow
Epoch 1/5
281/281 [==============================] - 2s 6ms/step - loss: 1.4060 - accuracy: 0.6912 - val_loss: 0.7704 - val_accuracy: 0.8241
Epoch 2/5
281/281 [==============================] - 1s 5ms/step - loss: 0.7699 - accuracy: 0.8233 - val_loss: 0.4825 - val_accuracy: 0.8938
Epoch 3/5
281/281 [==============================] - 1s 5ms/step - loss: 0.5459 - accuracy: 0.8676 - val_loss: 0.3345 - val_accuracy: 0.9226
Epoch 4/5
281/281 [==============================] - 2s 7ms/step - loss: 0.4149 - accuracy: 0.8969 - val_loss: 0.2489 - val_accuracy: 0.9422
Epoch 5/5
281/281 [==============================] - 1s 5ms/step - loss: 0.3370 - accuracy: 0.9162 - val_loss: 0.2054 - val_accuracy: 0.9489
1/1 [==============================] - 0s 79ms/step
cd 2023/04/25 16:45:32 WARNING mlflow.tensorflow: Unrecognized dataset type. Dataset logging skipped.
71/71 [==============================] - 0s 2ms/step - loss: 0.8731 - accuracy: 0.7983
Test score: 0.8730955719947815
Test accuracy: 0.7983080744743347
```

Reply:
Good catch. I've fixed this now and added the test case. From the docs https://www.tensorflow.org/api_docs/python/tf/keras/Model, `validation_data` can also be tensors, but we don't have `targets` for tensors yet. 

```
Loading data...
8982 train sequences
2246 test sequences
46 classes
Vectorizing sequence data...
x_train shape: (8982, 1000)
x_test shape: (2246, 1000)
Convert class vector to binary class matrix (for use with categorical_crossentropy)
y_train shape: (8982, 46)
y_test shape: (2246, 46)
Building model...
2023/04/25 21:26:20 INFO mlflow.utils.autologging_utils: Created MLflow autologging run with ID 'b3d7843db1974dd180cb7dea5018ae2e', which will track hyperparameters, performance metrics, model artifacts, and lineage information for the current tensorflow workflow
2023-04-25 21:26:20.454991: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
Epoch 1/5
  1/281 [..............................] - ETA: 1:32 - loss: 3.8544 - accuracy: 0.0312WARNING:tensorflow:Callback method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0015s vs `on_train_batch_end` time: 0.0199s). Check your callbacks.
281/281 [==============================] - 1s 3ms/step - loss: 1.3922 - accuracy: 0.6872 - val_loss: 0.7674 - val_accuracy: 0.8268
Epoch 2/5
281/281 [==============================] - 1s 2ms/step - loss: 0.7685 - accuracy: 0.8201 - val_loss: 0.4767 - val_accuracy: 0.8930
Epoch 3/5
281/281 [==============================] - 1s 2ms/step - loss: 0.5422 - accuracy: 0.8682 - val_loss: 0.3269 - val_accuracy: 0.9248
Epoch 4/5
281/281 [==============================] - 1s 2ms/step - loss: 0.4083 - accuracy: 0.8985 - val_loss: 0.2585 - val_accuracy: 0.9310
Epoch 5/5
281/281 [==============================] - 1s 2ms/step - loss: 0.3326 - accuracy: 0.9164 - val_loss: 0.1926 - val_accuracy: 0.9523
1/1 [==============================] - 0s 37ms/step
2023/04/25 21:26:28 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/prithvi.kannan/opt/miniconda3/envs/mlflow-dev-env/lib/python3.9/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils."
71/71 [==============================] - 0s 678us/step - loss: 0.8652 - accuracy: 0.7961
Test score: 0.8651795387268066
Test accuracy: 0.7960819005966187
```